src/Module/SubmitBookingOnPaymentHandler.php

```diff
@@ -0,0 +1,220 @@
+     *
+     * @since [*next-version*]
+     */
+    public function __invoke()
+    {
+        // Get the event instance and ensure it implements the interface
+        $event = func_get_arg(0);
+        if (!($event instanceof EventInterface)) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument is not an event instance'), null, null, $event
+            );
+        }
+
+        // Get the event params
+        $paymentId = $event->getParam(0);
+        $newStatus = $event->getParam(1);
+        $oldStatus = $event->getParam(2);
+
+        // Ensure payment was not already completed before. If it was, stop here.
+        if ($oldStatus === 'publish' || $oldStatus === 'complete') {
+            return;
+        }
+        // Check if the payment is being updated to a status of "completion". If not, stop here.
+        if ($newStatus !== 'publish' && $newStatus !== 'complete') {
+            return;
+        }
+
+        // Get the payment meta
+        $paymentMeta = \edd_get_payment_meta($paymentId);
+        // Get the items that were in the cart for this payment
+        $items = $this->_containerGet($paymentMeta, 'downloads');
+
+        // Get the cart item data keys from the cart item config
+        $dataKey       = $this->_containerGetPath($this->cartItemConfig, ['data', 'key']);
+        $eddBkKey      = $this->_containerGetPath($this->cartItemConfig, ['data', 'eddbk_key']);
+        $bookingIdKey  = $this->_containerGetPath($this->cartItemConfig, ['data', 'booking_id_key']);
+        $bookingIdPath = [$dataKey, $eddBkKey, $bookingIdKey];
+
+        // Alias expression builder
+        $b = $this->exprBuilder;
+
+        foreach ($items as $_item) {
+            try {
+                // Get the booking ID from the additional item data
+                $_bookingId = $this->_containerGetPath($_item, $bookingIdPath);
+                // Fetch the corresponding booking from storage
+                $_condition = $b->eq($b->var('id'), $b->lit($_bookingId));
+                $_bookings  = $this->bookingsSelectRm->select($_condition);
+                // Stop if no booking was found, or if multiple bookings matched the ID (for some reason?)
+                if ($this->_countIterable($_bookings) !== 1) {
+                    continue;
+                }
+
+                // Get the booking
+                $_booking = reset($_bookings);
+                $_booking = $this->transitioner->transition($_booking, T::TRANSITION_SUBMIT);
+
+                // Prepare the change set
+                $_changeSet = [
+                    'payment_id' => $paymentId,
+                    'client_id'  => \edd_get_payment_customer_id($paymentId),
+                    'status'     => $_booking->getStatus(),
+                ];
+
+                // Update booking
+                $this->bookingsUpdateRm->update($_changeSet, $_condition);
+            } catch (NotFoundExceptionInterface $exception) {
```

Comment:
This is only supposed to be triggered `_containerGet()` on ln. 191. The catch should immediately follow that line, to protect against runaway not-found-exceptions being accidentally thrown by other calls in the `try`.